File: src/daimon/face/web/src/bridge.js
// bridge.js — the only channel between the webview and the organ.
// Wraps pywebview's injected api (get_state / invoke) and the Python-pushed
// "daimon:state" event. The web layer holds no authority; every action is an
// action_id routed by the Python ActionRouter.

export const bridge = {
  async getState() {
    return window.pywebview.api.get_state();
  },
  async invoke(actionId, args = {}) {
    return window.pywebview.api.invoke(actionId, args);
  },
  // Python calls: window.dispatchEvent(new CustomEvent('daimon:state',{detail:<json>}))
  onState(cb) {
    window.addEventListener("daimon:state", (e) => cb(e.detail));
  },
};

File: src/daimon/face/bridge.py
"""The single typed JS<->Python surface exposed to the webview (pywebview js_api).
Exactly two methods: get_state (read serialized, non-secret view) and invoke
(route a known action_id through the shared ActionRouter). The web layer holds no
authority; this bridge cannot raise the ceiling and never returns secrets."""

from __future__ import annotations

from typing import Callable

from ..tray.actions import ActionRouter
from ..tray.state import TrayState
from .view_model import serialize


class FaceBridge:
    """js_api for the webview. `state_provider` reads a fresh TrayState (e.g.
    `daimon.tray.state.gather`); `router` is an ActionRouter over the real
    handlers (the same the AppKit menu uses)."""

    def __init__(self, router: ActionRouter, state_provider: Callable[[], TrayState]) -> None:
        self._router = router
        self._state = state_provider
        self._resizer: Callable[[int, int], None] | None = None

    def set_resizer(self, resizer: Callable[[int, int], None]) -> None:
        """Wire a window resizer (host-side) so the JS can fit the window to its
        content. Pure window mechanics — no authority, no state."""
        self._resizer = resizer

    def get_state(self) -> dict:
        return serialize(self._state())

    def invoke(self, action_id: str, args: dict | None = None) -> dict:
        res = self._router.dispatch(action_id)
        return {"ok": res.ok, "reason": res.reason}

    def resize_to(self, width: int, height: int) -> dict:
        if self._resizer is not None:
            self._resizer(int(width), int(height))
        return {"ok": True}
